components/org.wso2.carbon.dashboards.core/src/main/java/org/wso2/carbon/dashboards/core/internal/provider/impl/WidgetInfoProviderImpl.java

```diff
@@ -0,0 +1,158 @@
+        });
+        return metaInfoList;
+    }
+
+    @Override
+    public Optional<String> getWidgetConf(String widgetId) {
+        if (widgetMap.isEmpty()) {
+            initializeWidgetMap();
+        }
+        String widgetConfiguration = null;
+        if (widgetMap.containsKey(widgetId)) {
+            widgetConfiguration = readWidgetConf(widgetMap.get(widgetId));
+        } else {
+            log.error("Invalid widget ID !");
+        }
+        return Optional.ofNullable(widgetConfiguration);
+    }
+
+    @Override
+    public Optional<Path> getThumbnail(String widgetId) {
+        if (widgetMap.isEmpty()) {
+            initializeWidgetMap();
+        }
+        Path thumbnailPath = null;
+        if (widgetMap.containsKey(widgetId)) {
+            JsonObject widgetConf = jsonParser.parse(readWidgetConf(widgetMap.get(widgetId))).getAsJsonObject()
+                    .getAsJsonObject("info");
+            thumbnailPath = Paths
+                    .get(widgetMap.get(widgetId) + File.separator + widgetConf.get("thumbnail").getAsString());
+        } else {
+            log.error("Invalid widget ID ! No widget thumbnail !");
+        }
+        return Optional.ofNullable(thumbnailPath);
+    }
+
+    /**
+     * Check whether the given fragment/directory is a widget or not.
+     *
+     * @param widgetPath directory of the fragment
+     * @return true if the given fragment is a widget.
+     */
+    private boolean isWidget(Path widgetPath) {
+        return Files.exists(widgetPath.resolve(CONF_FILE), LinkOption.NOFOLLOW_LINKS);
+    }
+
+    /**
+     * This method is used to read the widget configuration file of the given widget.
+     *
+     * @param widgetPath Widget directory location
+     * @return widget configuration of the given widget.
+     */
+    private String readWidgetConf(Path widgetPath) {
+        Path widgetConfPath = widgetPath.resolve(CONF_FILE);
+        try {
+            return new String(Files.readAllBytes(widgetConfPath), StandardCharsets.UTF_8);
+        } catch (IOException ex) {
+            log.error("Error in reading widgetConf file in " + widgetPath.getFileName());
+        }
+        return null;
+    }
+
+    //TODO Since this method goes through all fragments in all apps and get widgets, need to rethink about the logic
+
+    /**
+     * Go through all the fragments in the uufapps directory and insert into widget map.
+     */
+    private void initializeWidgetMap() {
+        List<Path> uufAppsDirectories = null;
+        List<Path> componentDirectories = null;
+        List<Path> fragmentDirectories = null;
+        try {
+            uufAppsDirectories = Files.list(Paths.get(APP_PATH)).filter(Files::isDirectory)
+                    .collect(Collectors.toList());
+            for (Path path : uufAppsDirectories) {
+                componentDirectories = Files.list(path.resolve("components")).filter(Files::isDirectory)
```

Comment:
make these components and fragment keywords also constants and put it to a common constant class